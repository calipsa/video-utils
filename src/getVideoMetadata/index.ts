import { promises as fsPromises } from 'fs'

import * as tmp from 'tmp-promise'

import getMetadata from './getMetadata'

type Input = Buffer | string

export default async (input: Input) => {
  if (typeof input === 'string') {
    return getMetadata(input)
  }

  const { path, cleanup } = await tmp.file()
  try {
    await fsPromises.writeFile(path, input, 'binary')
    return await getMetadata(path)
  } finally {
    cleanup()
  }
}
