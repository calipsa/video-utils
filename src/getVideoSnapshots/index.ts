import * as tmp from 'tmp-promise'
import type { FfmpegCommand } from 'fluent-ffmpeg'

import type { Input } from '../types'
import writePolymorphic from '../utils/writePolymorphic'
// import noop from '../utils/noop'

import getImageBuffers from './getImageBuffers'

interface Options {
  input: Input,
  rate: number,
  format: 'jpg' | 'png',
  logError: (...args: any[]) => void,
  configureFfmpeg?: (ffmpegCommand: FfmpegCommand) => void,
}

export default async ({
  input,
  rate = 1,
  format = 'jpg',
  // logError = noop,
  configureFfmpeg,
}: Options) => {
  if (typeof input === 'string') {
    // eslint-disable-next-line no-return-await
    return await getImageBuffers(input, rate, format, configureFfmpeg)
  }

  const { path, cleanup } = await tmp.file()
  try {
    await writePolymorphic(path, input)
    return await getImageBuffers(path, rate, format, configureFfmpeg)
  } finally {
    await cleanup()
  }
}
