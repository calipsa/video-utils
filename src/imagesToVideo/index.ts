import { Writable } from 'stream'

import { ffmpeg } from 'ffmpeg-stream'

import type { PathOrReadable } from '../types'
import streamImages from './streamImages'

interface Params {
  images: PathOrReadable[],
  fps: number,
}

export default ({
  images,
  fps,
}: Params) => {
  const converter = ffmpeg()

  const input = converter.input({
    f: 'image2pipe',
    r: fps,
  }) as Writable

  // const output = converter.output({
  //   f: 'mpegts',
  //   vcodec: 'mpeg2video',
  //   tune: 'zerolatency',
  //   preset: 'ultrafast',
  //   'q:v': 1,
  // })

  const output = converter.output({
    f: 'avi',
    vcodec: 'wmv2',
    // buffer: true,
  })

  streamImages(images, input)
  converter.run()

  return output
}
