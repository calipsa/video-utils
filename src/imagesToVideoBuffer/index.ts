import type { Input } from '../types'

import convert from './convert'

interface Params {
  images: Input[],
  fps: number,
}

export default ({
  images,
  fps,
}: Params) => convert(images, fps)
