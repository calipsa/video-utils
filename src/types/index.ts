import { Readable } from 'stream'

export type PathOrReadable = string | Readable
export type Input = PathOrReadable | Buffer
