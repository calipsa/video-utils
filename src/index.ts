import getVideoMetadata from './getVideoMetadata'
import getVideoSnapshots from './getVideoSnapshots'
import imagesToVideoStream from './imagesToVideoStream'
import imagesToVideoBuffer from './imagesToVideoBuffer'

export {
  getVideoMetadata,
  getVideoSnapshots,
  imagesToVideoStream,
  imagesToVideoBuffer,
}
