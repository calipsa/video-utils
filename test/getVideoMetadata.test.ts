import fs from 'fs'
// const P = require('path')
import { streamToBuffer } from '@calipsa/stream-utils'
import {
  identity,
  flow,
  omit,
} from 'lodash'

import { getVideoMetadata } from '../src'
import getHash from './utils/getHash'

const files = [
  'building.mp4',
  'building.wmv',
]

const pathToInput = {
  string: identity,
  buffer: flow(fs.createReadStream, streamToBuffer),
} as const

async function getHashes(path, type) {
  const f = pathToInput[type]
  const input = await f(path)
  const metadata = await getVideoMetadata(input)
  return omit(metadata, [
    'format.filename',
  ])
}

describe('Get video metdata', () => {
  beforeEach(() => {
    jest.setTimeout(120000)
  })

  for (const file of files) {
    it(`Should get metadata (${file})`, async () => {
      console.log(file)
      const path = `${__dirname}/videos/${file}`
      const types = Object.keys(pathToInput)
      const metadataPromises = types.map(type => getHashes(path, type))
      const o = {}
      for (let i = 0; i < types.length; ++i) {
        const type = types[i]
        const p = metadataPromises[i]
        o[type] = await p
        if (i > 0) {
          expect(o[type]).toEqual(o[types[0]])
        }
      }
      expect(o).toMatchSnapshot()
    })
  }
})
