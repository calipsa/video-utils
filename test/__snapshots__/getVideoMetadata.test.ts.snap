// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get video metdata Should get metadata (building.mp4) 1`] = `
Object {
  "buffer": Object {
    "chapters": Array [],
    "format": Object {
      "bit_rate": 2876684,
      "duration": 14.6,
      "format_long_name": "QuickTime / MOV",
      "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
      "nb_programs": 0,
      "nb_streams": 1,
      "probe_score": 100,
      "size": 5249950,
      "start_time": 0,
      "tags": Object {
        "compatible_brands": "mp42mp41isomavc1",
        "creation_time": "2019-02-19T10:21:11.000000Z",
        "major_brand": "mp42",
        "minor_version": "0",
      },
    },
    "streams": Array [
      Object {
        "avg_frame_rate": "30/1",
        "bit_rate": 2873072,
        "bits_per_raw_sample": 8,
        "chroma_location": "left",
        "closed_captions": 0,
        "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
        "codec_name": "h264",
        "codec_tag": "0x31637661",
        "codec_tag_string": "avc1",
        "codec_time_base": "1/60",
        "codec_type": "video",
        "coded_height": 720,
        "coded_width": 1280,
        "color_primaries": "bt709",
        "color_range": "tv",
        "color_space": "bt709",
        "color_transfer": "bt709",
        "display_aspect_ratio": "N/A",
        "disposition": Object {
          "attached_pic": 0,
          "clean_effects": 0,
          "comment": 0,
          "default": 1,
          "dub": 0,
          "forced": 0,
          "hearing_impaired": 0,
          "karaoke": 0,
          "lyrics": 0,
          "original": 0,
          "timed_thumbnails": 0,
          "visual_impaired": 0,
        },
        "duration": 14.6,
        "duration_ts": 438,
        "field_order": "unknown",
        "has_b_frames": 2,
        "height": 720,
        "id": "N/A",
        "index": 0,
        "is_avc": "true",
        "level": 32,
        "max_bit_rate": "N/A",
        "nal_length_size": 4,
        "nb_frames": 438,
        "nb_read_frames": "N/A",
        "nb_read_packets": "N/A",
        "pix_fmt": "yuv420p",
        "profile": "High",
        "r_frame_rate": "30/1",
        "refs": 1,
        "sample_aspect_ratio": "N/A",
        "start_pts": 0,
        "start_time": 0,
        "tags": Object {
          "creation_time": "2019-02-19T10:21:11.000000Z",
          "encoder": "AVC Coding",
          "handler_name": "L-SMASH Video Handler",
          "language": "und",
        },
        "time_base": "1/30",
        "timecode": "N/A",
        "width": 1280,
      },
    ],
  },
  "string": Object {
    "chapters": Array [],
    "format": Object {
      "bit_rate": 2876684,
      "duration": 14.6,
      "format_long_name": "QuickTime / MOV",
      "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
      "nb_programs": 0,
      "nb_streams": 1,
      "probe_score": 100,
      "size": 5249950,
      "start_time": 0,
      "tags": Object {
        "compatible_brands": "mp42mp41isomavc1",
        "creation_time": "2019-02-19T10:21:11.000000Z",
        "major_brand": "mp42",
        "minor_version": "0",
      },
    },
    "streams": Array [
      Object {
        "avg_frame_rate": "30/1",
        "bit_rate": 2873072,
        "bits_per_raw_sample": 8,
        "chroma_location": "left",
        "closed_captions": 0,
        "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
        "codec_name": "h264",
        "codec_tag": "0x31637661",
        "codec_tag_string": "avc1",
        "codec_time_base": "1/60",
        "codec_type": "video",
        "coded_height": 720,
        "coded_width": 1280,
        "color_primaries": "bt709",
        "color_range": "tv",
        "color_space": "bt709",
        "color_transfer": "bt709",
        "display_aspect_ratio": "N/A",
        "disposition": Object {
          "attached_pic": 0,
          "clean_effects": 0,
          "comment": 0,
          "default": 1,
          "dub": 0,
          "forced": 0,
          "hearing_impaired": 0,
          "karaoke": 0,
          "lyrics": 0,
          "original": 0,
          "timed_thumbnails": 0,
          "visual_impaired": 0,
        },
        "duration": 14.6,
        "duration_ts": 438,
        "field_order": "unknown",
        "has_b_frames": 2,
        "height": 720,
        "id": "N/A",
        "index": 0,
        "is_avc": "true",
        "level": 32,
        "max_bit_rate": "N/A",
        "nal_length_size": 4,
        "nb_frames": 438,
        "nb_read_frames": "N/A",
        "nb_read_packets": "N/A",
        "pix_fmt": "yuv420p",
        "profile": "High",
        "r_frame_rate": "30/1",
        "refs": 1,
        "sample_aspect_ratio": "N/A",
        "start_pts": 0,
        "start_time": 0,
        "tags": Object {
          "creation_time": "2019-02-19T10:21:11.000000Z",
          "encoder": "AVC Coding",
          "handler_name": "L-SMASH Video Handler",
          "language": "und",
        },
        "time_base": "1/30",
        "timecode": "N/A",
        "width": 1280,
      },
    ],
  },
}
`;

exports[`Get video metdata Should get metadata (building.wmv) 1`] = `
Object {
  "buffer": Object {
    "chapters": Array [],
    "format": Object {
      "bit_rate": 5520248,
      "duration": 14.6,
      "format_long_name": "ASF (Advanced / Active Streaming Format)",
      "format_name": "asf",
      "nb_programs": 0,
      "nb_streams": 1,
      "probe_score": 100,
      "size": 10074453,
      "start_time": 0,
      "tags": Object {
        "compatible_brands": "mp42mp41isomavc1",
        "encoder": "Lavf58.26.101",
        "major_brand": "mp42",
        "minor_version": "0",
      },
    },
    "streams": Array [
      Object {
        "avg_frame_rate": "30/1",
        "bit_rate": "N/A",
        "bits_per_raw_sample": "N/A",
        "chroma_location": "unspecified",
        "closed_captions": 0,
        "codec_long_name": "Windows Media Video 8",
        "codec_name": "wmv2",
        "codec_tag": "0x32564d57",
        "codec_tag_string": "WMV2",
        "codec_time_base": "1/30",
        "codec_type": "video",
        "coded_height": 720,
        "coded_width": 1280,
        "color_primaries": "unknown",
        "color_range": "unknown",
        "color_space": "unknown",
        "color_transfer": "unknown",
        "display_aspect_ratio": "N/A",
        "disposition": Object {
          "attached_pic": 0,
          "clean_effects": 0,
          "comment": 0,
          "default": 0,
          "dub": 0,
          "forced": 0,
          "hearing_impaired": 0,
          "karaoke": 0,
          "lyrics": 0,
          "original": 0,
          "timed_thumbnails": 0,
          "visual_impaired": 0,
        },
        "duration": 14.6,
        "duration_ts": 14600,
        "field_order": "unknown",
        "has_b_frames": 0,
        "height": 720,
        "id": "N/A",
        "index": 0,
        "level": "-99",
        "max_bit_rate": "N/A",
        "nb_frames": "N/A",
        "nb_read_frames": "N/A",
        "nb_read_packets": "N/A",
        "pix_fmt": "yuv420p",
        "profile": "unknown",
        "r_frame_rate": "30/1",
        "refs": 1,
        "sample_aspect_ratio": "N/A",
        "start_pts": 0,
        "start_time": 0,
        "time_base": "1/1000",
        "timecode": "N/A",
        "width": 1280,
      },
    ],
  },
  "string": Object {
    "chapters": Array [],
    "format": Object {
      "bit_rate": 5520248,
      "duration": 14.6,
      "format_long_name": "ASF (Advanced / Active Streaming Format)",
      "format_name": "asf",
      "nb_programs": 0,
      "nb_streams": 1,
      "probe_score": 100,
      "size": 10074453,
      "start_time": 0,
      "tags": Object {
        "compatible_brands": "mp42mp41isomavc1",
        "encoder": "Lavf58.26.101",
        "major_brand": "mp42",
        "minor_version": "0",
      },
    },
    "streams": Array [
      Object {
        "avg_frame_rate": "30/1",
        "bit_rate": "N/A",
        "bits_per_raw_sample": "N/A",
        "chroma_location": "unspecified",
        "closed_captions": 0,
        "codec_long_name": "Windows Media Video 8",
        "codec_name": "wmv2",
        "codec_tag": "0x32564d57",
        "codec_tag_string": "WMV2",
        "codec_time_base": "1/30",
        "codec_type": "video",
        "coded_height": 720,
        "coded_width": 1280,
        "color_primaries": "unknown",
        "color_range": "unknown",
        "color_space": "unknown",
        "color_transfer": "unknown",
        "display_aspect_ratio": "N/A",
        "disposition": Object {
          "attached_pic": 0,
          "clean_effects": 0,
          "comment": 0,
          "default": 0,
          "dub": 0,
          "forced": 0,
          "hearing_impaired": 0,
          "karaoke": 0,
          "lyrics": 0,
          "original": 0,
          "timed_thumbnails": 0,
          "visual_impaired": 0,
        },
        "duration": 14.6,
        "duration_ts": 14600,
        "field_order": "unknown",
        "has_b_frames": 0,
        "height": 720,
        "id": "N/A",
        "index": 0,
        "level": "-99",
        "max_bit_rate": "N/A",
        "nb_frames": "N/A",
        "nb_read_frames": "N/A",
        "nb_read_packets": "N/A",
        "pix_fmt": "yuv420p",
        "profile": "unknown",
        "r_frame_rate": "30/1",
        "refs": 1,
        "sample_aspect_ratio": "N/A",
        "start_pts": 0,
        "start_time": 0,
        "time_base": "1/1000",
        "timecode": "N/A",
        "width": 1280,
      },
    ],
  },
}
`;
